subsys/mpsl/clock_ctrl/mpsl_clock_ctrl.c

```diff
@@ -0,0 +1,360 @@
+ *
+ * The callback function provided to clock control to notify about LFCLK request being finished.
+ */
+static void lfclk_request_cb(struct onoff_manager *mgr, struct onoff_client *cli, uint32_t state,
+			     int res)
+{
+	struct clock_onoff_state *clock_state = CONTAINER_OF(cli, struct clock_onoff_state, cli);
+
+	clock_state->clk_req_rsp = res;
+	k_sem_give(&clock_state->sem);
+}
+
+/** @brief Wait for LFCLK to be ready.
+ *
+ * The function can time out if there is no response from clock control drvier until
+ * MPSL_LFCLK_REQUEST_WAIT_TIMEOUT_MS.
+ *
+ * @note For nRF54H SoC series waiting for LFCLK can't block the system work queue. The nrf2 clock
+ *       control driver can return -TIMEDOUT due not handled response from sysctrl.
+ *
+ * @retval 0 LFCLK is ready.
+ * @retval -NRF_EINVAL There were no LFCLK request.
+ * @retval -NRF_EFAULT LFCLK request failed.
+ */
+static int32_t m_lfclk_wait(void)
+{
+	int32_t err;
+
+	if (atomic_get(&m_lfclk_state.m_clk_ready) == (atomic_val_t) true) {
+		return 0;
+	}
+
+	/* Check if lfclk has been requested */
+	if (atomic_get(&m_lfclk_state.m_clk_refcnt) <= (atomic_val_t)0) {
+		return -NRF_EINVAL;
+	}
+
+	/* Wait for response from clock control */
+	err = k_sem_take(&m_lfclk_state.sem, K_MSEC(MPSL_LFCLK_REQUEST_WAIT_TIMEOUT_MS));
+	if (err < 0) {
+		/* Do a gracefull cancel of the request, the function release does this
+		 * as well as and relase.
+		 */
+		(void)m_lfclk_release();
+
+		return -NRF_EFAULT;
+	}
+
+	if (IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF2) && m_lfclk_state.clk_req_rsp == -ETIMEDOUT) {
+		/* Due to NCSDK-31169, temporarily allow for LFCLK request to timeout.
+		 * That doens't break anything now because the LFCLK requested clock is
+		 * 500PPM and such LFCLK should be running from boot of the radio core.
+		 */
+		LOG_WRN("LFCLK could not be started: %d", m_lfclk_state.clk_req_rsp);
+		return 0;
+	} else if (m_lfclk_state.clk_req_rsp < 0) {
+		__ASSERT(false, "LFCLK could not be started, reason: %d",
+			 m_lfclk_state.clk_req_rsp);
+		/* Possible failure reasons:
+		 *  # -ERRTIMEDOUT - nRFS service timeout
+		 *  # -EIO - nRFS service error
+		 *  # -ENXIO - request rejected
+		 * All these mean failure for MPLS.
+		 */
+		return -NRF_EFAULT;
+	}
+
+	atomic_set(&m_lfclk_state.m_clk_ready, (atomic_val_t) true);
+
+	return 0;
+}
+
+#if defined(CONFIG_CLOCK_CONTROL_NRF)
+
+static void m_hfclk_request(void)
+{
+	/* The z_nrf_clock_bt_ctlr_hf_request doesn't count references to HFCLK,
+	 * it is caller responsibility handle requests and releases counting.
+	 */
+	if (atomic_inc(&m_hfclk_refcnt) > (atomic_val_t)0) {
+		return;
+	}
+
+	z_nrf_clock_bt_ctlr_hf_request();
+}
+
+static void m_hfclk_release(void)
+{
+	/* The z_nrf_clock_bt_ctlr_hf_request doesn't count references to HFCLK,
+	 * it is caller responsibility to not release the clock if there is
+	 * other request pending.
+	 */
+	if (atomic_get(&m_hfclk_refcnt) < (atomic_val_t)1) {
+		return;
+	}
```

Comment:
Should we have a warning here?
```suggestion
	if (atomic_get(&m_hfclk_refcnt) < (atomic_val_t)1) {
		LOG_WRN("Mismatch between HFCLK request/release");
		return;
	}
```
Similar for the other implementation of `m_hfclk_release()`

Reply:
Good idea, thank you!

Reply:
Added to both implementation of the function for `nrf/nrf2 clock control`